src/c_api.cpp

```diff
@@ -364,9 +373,13 @@ class Booster {
  private:
   const Dataset* train_data_;
   std::unique_ptr<Boosting> boosting_;
-  std::unique_ptr<Predictor> single_row_predictor_;
-  PredictFunction single_row_predict_function_;
-  int64_t single_row_num_pred_in_one_row_;
+  std::unique_ptr<Predictor> single_row_predictor_[PREDICTOR_TYPES];
+  bool single_row_predictor_pred_early_stop[PREDICTOR_TYPES];
+  int single_row_predictor_pred_early_stop_freq[PREDICTOR_TYPES];
+  double single_row_predictor_pred_early_stop_margin[PREDICTOR_TYPES];
+  int single_row_predictor_iter_[PREDICTOR_TYPES];
```

Comment:
the config has a lot of info, but it seems I only care about these values (early_stop, early_stop_freq, early_stop_margin, iter), so I only cache them in an array

Reply:
+1 on the approach